### PR TITLE
Ledger: Cache the result of getValidators

### PR DESCRIPTION
**Pull request by Geod24**
```
Most calls to `getValidators` are done for the current (or next) height,
and historical data are rarely looked up.
We can take advantage of this by caching based on the height,
to reduce the number of memory allocations or database lookups.
```

I'm a bit on the fence about this, since it can defeat some expectations (as seen in `addPreimages` invalidation).
But we call this method a lot and either we need to start passing the array around, or caching like here.

### CODE REVIEW

**Review by linked0 — APPROVED** [review]
Nice!

File: source/agora/consensus/Ledger.d [hunk 1]
```diff
@@ -116,6 +116,9 @@ public class Ledger
     /// but less than current time + block_time_offset_tolerance
     public Duration block_time_offset_tolerance;
 
+    /// Cache the result of `getValidators` to allow frequent calls
+    protected HeightCache!(ValidatorInfo[]) validatorCache;
```
Comment (omerfirmak) [inline source/agora/consensus/Ledger.d:120]: Wouldnt it be a less intrusive and more intuitive to have ValidatorSet to do the caching, stuff like addPreimage would be easier to deal with too.

Comment (Geod24) [reply source/agora/consensus/Ledger.d:120]: Yeah that makes much more sense 🤔

**Comment by Geod24** [comment]
Superseded by #3025

**Pull request closed without merging**